PyInstaller/loader/pyimod03_importers.py

```diff
@@ -491,6 +491,119 @@ def get_filename(self, fullname):
+
+        if filename is None:
+            trace("# %s not found in PYZ", fullname)
+            return None
+
+        is_pkg, bytecode = self._pyz_archive.extract(filename)
+        if is_pkg:
+            origin = pyi_os_path.os_path_join(pyi_os_path.os_path_join(
+                SYS_PREFIX, filename.replace('.', pyi_os_path.os_sep)), '__init__.pyc')
+        else:
+            origin = pyi_os_path.os_path_join(SYS_PREFIX, filename.replace('.', pyi_os_path.os_sep) + '.pyc')
+
+        if not hasattr(self, 'module_code'):
+            self.module_code = {}
+
+        self.module_code[fullname] = bytecode
+        return _frozen_importlib.ModuleSpec(fullname, self, is_package=is_pkg, origin=origin)
+
+    def create_module(self, spec):
+        """
+        PEP-451 loader.create_module() method for the ``sys.meta_path`` hook.
+
+        Loaders may also implement create_module() that will return a new module to exec. It may return None to
+        indicate that the default module creation code should be used. One use case, though atypical, for
+        create_module() is to provide a module that is a subclass of the builtin module type. Most loaders will not
+        need to implement create_module(),
+
+        create_module() should properly handle the case where it is called more than once for the same spec/module.
+        This may include returning None or raising ImportError.
+        """
+        return imp_new_module(spec.name)
+
+    def exec_module(self, module):
+        """
+        PEP-451 loader.exec_module() method for the ``sys.meta_path`` hook.
+
+        Loaders will have a new method, exec_module(). Its only job is to "exec" the module and consequently
+        populate the module's namespace. It is not responsible for creating or preparing the module object, nor
+        for any cleanup afterward. It has no return value. exec_module() will be used during both loading and
+        reloading.
+
+        exec_module() should properly handle the case where it is called more than once. For some kinds of modules
+        this may mean raising ImportError every time after the first time the method is called. This is
+        particularly relevant for reloading, where some kinds of modules do not support in-place reloading.
+        """
+        if hasattr(self, 'module_code') and module.__spec__.name in self.module_code:
+            bytecode = self.module_code[module.__spec__.name]
+            del self.module_code[module.__spec__.name]
+        else:
+            bytecode = self.get_code(module.__spec__.name)
+
+        if not hasattr(module, '__file__') and module.__spec__.origin:
+            module.__file__ = module.__spec__.origin
+
+        ### Set __path__  if 'fullname' is a package.
+        # Python has modules and packages. A Python package is container
+        # for several modules or packages.
+        if hasattr(module, '__file__') and module.__spec__.submodule_search_locations is not None:
```

Comment:
I should clarify: you _cannot_ check truthiness here because `submodule_search_locations` may be empty.